text/0000-expand-math.md

```diff
@@ -0,0 +1,230 @@
+
+## Numeric Hierarchy
+
+Current the Pony numeric type hierarchy is as follows:
+
+```mermaid
+classDiagram
+class Any
+<<interface>> Any
+class Real
+<<trait>> Real
+class FloatingPoint
+<<trait>> FloatingPoint
+class Integer
+<<trait>> Integer
+class SignedInteger
+<<trait>> SignedInteger
+class UnsignedInteger
+<<trait>> UnsignedInteger
+
+Any <-- Real
+Real <-- FloatingPoint
+Real <-- Integer
+Integer <-- SignedInteger
+Integer <-- UnsignedInteger
+
+FloatingPoint <-- F32
+FloatingPoint <-- F64
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+This RFC introduces four more numeric types: `Rational`, `BigInt`, `BigFloat`, and `Complex`. These fit into the numeric type hierarchy in the following manner:
+
+```mermaid
+classDiagram
+class Any
+<<interface>> Any
+class Real
+<<trait>> Real
+class FloatingPoint
+<<trait>> FloatingPoint
+class Integer
+<<trait>> Integer
+class SignedInteger
+<<trait>> SignedInteger
+class UnsignedInteger
+<<trait>> UnsignedInteger
+
+Any <-- Real
+Any <-- Complex~Real~
+Real <-- FloatingPoint
+Real <-- Integer
+Real <-- Rational~Integer~
+Integer <-- SignedInteger
+Integer <-- UnsignedInteger
+Integer <-- BigInt
+
+FloatingPoint <-- F32
+FloatingPoint <-- F64
+FloatingPoint <-- BigFloat
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+## Methods of Concern
+
+```pony
+trait val Real[A: Real[A] val] is
+  (Stringable & _ArithmeticConvertible & Comparable[A])
+  ...
+  new val min_value()
+  new val max_value()
+  ...
+```
+
+`Real` will exist above `Rational`, `BigInt`, and `BigFloat` and as such would require defining the above methods for these types. `Rational` can be defined as minimum and maximum of the numerator, however `BigInt` and `BigFloat` by definition have arbitrary precision making defining a minimum and maximum difficult at the least -- if we define them as the minimum and maximum of a machine-sized int and float, or define them as -Inf and Inf -- or impossible at the worst -- if we define them by their possible limits which are arbitrary.
+
+```pony
+trait val Integer[A: Integer[A] val] is Real[A]
+  ...
+  fun op_and(y: A): A => this and y
+  fun op_or(y: A): A => this or y
+  fun op_xor(y: A): A => this xor y
+  fun op_not(): A => not this
```

Comment:
This can't be implemented by a BigInt, so it's odd that a BigInt can't actually be an Integer

Reply:
Sorry, I don't understand your comment. Can you try explaining in a different way?

Reply:
You can't do a bitwise "not" operation on a `BigInt` - at least not by the normal semantics of a bitwise "not" operation.

You can't because the you'd expect that any bits "more significant" than the most significant bit of the value would be set to `1`.

But a `BigInt` has no upper bound, and thus there is no limit to the number of leading `1` bits implied by inverting its bits

Reply:
I still don't understand @jemc.

> "This can't be implemented by a BigInt, so it's odd that a BigInt can't actually be an Integer"

Can you explain how what you said is related to Jason's comment?

I don't understand why not being able to be implemented BigInt makes it odd that BigInt can't actually be an Integer.
I don't understand what is being said here.

Reply:
I'll copy Jason's words and insert parentheticals with my own commentary:

This (the `op_not` function) can't be implemented by a `BigInt` (for the reason mentioned in my previous comment)

So it's odd that a `BigInt` (which is conceptually a kind of integer) can't actually be an `Integer` (it cannot be a subtype of the `Integer` trait, because the `Integer` trait includes the `op_not` function).

Reply:
Thanks @jemc. I understand now.